### PR TITLE
Feature/update user metadata

### PR DESCRIPTION
**Pull request by garyluu**
For #1370 

Fix for user profiles not being sync'd upon token creation.

Add tests for tokenResource

### CODE REVIEW

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TokenResource.java [hunk 1]
```diff
@@ -142,11 +138,23 @@ public TokenResource(TokenDAO tokenDAO, UserDAO enduserDAO, HttpClient client, C
         this.cachingAuthenticator = cachingAuthenticator;
     }
 
+    public static Token createGoogleToken(String accessToken, String refreshToken, long userID, String googleLoginName) {
```
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TokenResource.java]: ~Seems like this can be private~

That said, this looks like just a constructor in disguise. 
Put it in Token and then it would be easier to find and re-use.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TokenResource.java [hunk 2]
```diff
@@ -301,135 +308,150 @@ public Token addToken(@ApiParam("code") String satellizerJson) {
...
+    }
+
+    /**
+     * Gets the Google TokenResponse
+     *
+     * @param code        The satellizer code
+     * @param redirectUri The Google redirectUri
+     * @return
+     */
+    public TokenResponse getTokenResponse(String code, String redirectUri) {
```
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TokenResource.java]: looks private

File: dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        return fakeTokenResponse;
+    }
+
+    private static Userinfoplus getFakeUserinfoplus() {
+        Userinfoplus fakeUserinfoplus = new Userinfoplus();
+        fakeUserinfoplus.setEmail("potato@gmail.com");
+        fakeUserinfoplus.setGivenName("Beef");
+        fakeUserinfoplus.setFamilyName("Stew");
+        fakeUserinfoplus.setGender("New classification");
+        fakeUserinfoplus.setPicture("https://i.kym-cdn.com/photos/images/newsfeed/001/207/210/b22.jpg");
```
Comment (denis-yuen) [inline dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java]: Change this. not really appropriate

File: dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        fakeToken.setUserId(1);
+        fakeToken.setUsername("fakeTokenUsername");
+        return fakeToken;
+    }
+
+    private static User getFakeUser() {
+        User fakeUser = new User();
+        fakeUser.setUsername("fakeUser");
+        fakeUser.setIsAdmin(false);
+        fakeUser.setAvatarUrl("https://i.kym-cdn.com/photos/images/newsfeed/001/207/210/b22.jpg");
```
Comment (denis-yuen) [inline dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java]: Ditto

File: dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+/**
+ * @author gluu
+ * @since 24/07/18
+ */
+public class TokenResourceTest {
+    private static final TokenDAO tokenDAO = mock(TokenDAO.class);
```
Comment (denis-yuen) [inline dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java]: Seems like a miscommunication here, I meant that any interaction with Google can be mocked. 
I'd prefer not mocking the DAOs since we use a live database in our integration tests anyway.
Sorry

File: dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        fakeUser.setUsername("fakeUser");
+        fakeUser.setIsAdmin(false);
+        fakeUser.setAvatarUrl("https://i.kym-cdn.com/photos/images/newsfeed/001/207/210/b22.jpg");
+        fakeUser.setCurator(false);
+        return fakeUser;
+    }
+
+    @Before
+    public void setup() {
+        doReturn(getFakeTokenResponse()).when(tokenResourceSpy).getTokenResponse(anyString(), anyString());
```
Comment (denis-yuen) [inline dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java]: These two lines are ok for example.

File: dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        doReturn(getFakeUserinfoplus()).when(tokenResourceSpy).getUserInfo(anyString());
+        doReturn(1l).when(tokenDAO).create(any());
```
Comment (denis-yuen) [inline dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java]: These three lines seem unnecessary, please test against the actual database.

File: dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+    @After
+    public void tearDown() {
+        reset(tokenDAO);
+        reset(enduserDAO);
+        reset(tokenResourceSpy);
+    }
+
+    /**
+     * For a non-existing user, checks that two tokens (Dockstore and Google) were created
```
Comment (denis-yuen) [inline dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java]: This checks on a dockstore token, it doesn't seem to check on a google token?

File: dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+            Assert.assertEquals("fakeContent", token.getContent());
+            Assert.assertEquals("fakeTokenUsername", token.getUsername());
+            Assert.assertEquals(TokenType.DOCKSTORE, token.getTokenSource());
+        } catch (IOException e) {
+            Assert.fail("Mapper problem: " + e.getMessage());
+        }
+    }
+
+    /**
+     * For an existing user without a Google token, checks that a token (Google) was created exactly once.
```
Comment (denis-yuen) [inline dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java]: Seems to check that a dockstore token was created?

Comment (garyluu) [reply dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java]: Checks that some token was created.  Checks that the Dockstore token is returned (GitHub endpoint also returns the Dockstore token)

File: dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java (hunk 8 above)
Comment (garyluu) [reply dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java]: Checks that two tokens were created. Checks that the Dockstore token is returned (GitHub endpoint also returns the Dockstore token)

File: dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java (hunk 9 above)
Comment (denis-yuen) [reply dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java]: Ah ok, let's match the comment and check that the token created is actually a google token

File: dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java (hunk 8 above)
Comment (denis-yuen) [reply dockstore-webservice/src/test/java/io/dockstore/webservice/TokenResourceTest.java]: Ah ok, please check that one of each is created

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TokenResource.java [hunk 10]
```diff
@@ -285,9 +282,8 @@ public Token addGitlabToken(@ApiParam(hidden = true) @Auth User user, @QueryPara
     @Timed
     @UnitOfWork
     @Path("/github")
-    @ApiOperation(value = "Allow satellizer to post a new GitHub token to dockstore", authorizations = { @Authorization(value = JWT_SECURITY_DEFINITION_NAME) },
-        notes = "A post method is required by saetillizer to send the GitHub token",
-        response = Token.class)
+    @ApiOperation(value = "Allow satellizer to post a new GitHub token to dockstore", authorizations = {
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TokenResource.java]: I'd mentioned this in a private Slack, but mentioning satellizer sounds weird to me. Anybody could post a token, not just the Satellizer library, can't they? It seems odd to single out a specific library in our Swagger doc. Not really part of this PR, but seeing it again triggered that thought. :)

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TokenResource.java [hunk 11]
```diff
@@ -301,135 +297,150 @@ public Token addToken(@ApiParam("code") String satellizerJson) {
...
     public Token addGoogleToken(@ApiParam(hidden = true) @Auth Optional<User> authUser, @ApiParam("code") String satellizerJson) {
         Gson gson = new Gson();
         JsonElement element = gson.fromJson(satellizerJson, JsonElement.class);
         JsonObject satellizerObject = element.getAsJsonObject();
-
         final String code = satellizerObject.get("code").getAsString();
         final String redirectUri = satellizerObject.get("redirectUri").getAsString();
         String accessToken = null;
         String refreshToken = null;
         if (accessToken == null) {
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TokenResource.java]: accessToken is set to null two lines up, no need to check for null again (not changed, but noticed in trying to understand code).

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TokenResource.java (hunk 10 above)
Comment (denis-yuen) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TokenResource.java]: Hah, was just working on swagger examples in the TRS. 
I don't think there's anything wrong in providing specific examples of how the endpoint is used but it might be worth noting that this can probably be re-used by other libraries.

**Comment by garyluu** [comment]
Created #1660 for follow-up

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GoogleHelper.java [hunk 12]
```diff
@@ -152,7 +160,30 @@ static boolean isValidAudience(Tokeninfo tokenInfo) {
...
+        final AuthorizationCodeFlow flow = new AuthorizationCodeFlow.Builder(BearerToken.authorizationHeaderAccessMethod(), TokenResource.HTTP_TRANSPORT,
+            TokenResource.JSON_FACTORY, new GenericUrl(GoogleHelper.GOOGLE_ENCODED_URL),
+            new ClientParametersAuthentication(googleClientID, googleClientSecret), googleClientID,
+            GoogleHelper.GOOGLE_AUTHORIZATION_SERVICE_ENCODED_URL).build();
+        try {
+            return flow.newTokenRequest(code).setRedirectUri(redirectUri)
+                .setRequestInitializer(request -> request.getHeaders().setAccept("application/json")).execute();
+        } catch (IOException e) {
+            LOG.error("Retrieving accessToken was unsuccessful");
+            throw new CustomWebApplicationException("Could not retrieve google.com token based on code", HttpStatus.SC_BAD_REQUEST);
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GoogleHelper.java]: This is super minor, but I would not say `google.com`, as you will typically be logging in with another domain, e.g., I log in with a ucsc.edu domain. Maybe just `google token`.
